clang/lib/Headers/cheriintrin.h

```diff
@@ -0,0 +1,132 @@
+/*-------------- cheriintrin.h - CHERI intrinsics -----------------------
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2018-2020 Alex Richardson
+ * All rights reserved.
+ *
+ * This software was developed by SRI International and the University of
+ * Cambridge Computer Laboratory under DARPA/AFRL contract FA8750-10-C-0237
+ * ("CTSRD"), as part of the DARPA CRASH research programme.
+ *
+ * This software was developed by SRI International and the University of
+ * Cambridge Computer Laboratory (Department of Computer Science and
+ * Technology) under DARPA contract HR0011-18-C-0016 ("ECATS"), as part of the
+ * DARPA SSITH research programme.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#ifndef __CHERIINTRIN_H
+#define __CHERIINTRIN_H
+
+#if !__has_feature(capabilities)
+#error "<cheriintrin.h> should only be included when CHERI is supported"
+#endif
+
+/*
+ * Provide shorter names for CHERI builtins that are portable across all
+ * CHERI-enabled architectures:
+ */
+#define cheri_getaddress(x) __builtin_cheri_address_get(x)
+#define cheri_setaddress(x, y) __builtin_cheri_address_set((x), (y))
+#define cheri_getbase(x) __builtin_cheri_base_get(x)
+#define cheri_getlength(x) __builtin_cheri_length_get(x)
+#define cheri_getoffset(x) __builtin_cheri_offset_get(x)
+#define cheri_incoffset(x, y) __builtin_cheri_offset_increment((x), (y))
```

Comment:
Is this a meaningful thing to provide, given it's just pointer arithmetic that turns into a GEP and there isn't even an underlying intrinsic any more? Yes, it's another way to be explicit about provenance, but I think we should encourage the use of plain `+` with suitable casts/annotations rather than what looks like a potentially magic function?

Reply:
It looks like we've only got one use of `cheri_incoffset` in CheriBSD and that's in swap page-in code that should be replaced with `cheri_buildcap`.